L1Trigger/TrackerTFP/src/GeometricProcessor.cc

```diff
@@ -0,0 +1,128 @@
+#include "L1Trigger/TrackerTFP/interface/GeometricProcessor.h"
+
+#include <numeric>
+#include <algorithm>
+#include <iterator>
+#include <deque>
+#include <vector>
+
+using namespace std;
+using namespace edm;
+using namespace trackerDTC;
+
+namespace trackerTFP {
+
+  GeometricProcessor::GeometricProcessor(const ParameterSet& iConfig, const Setup* setup, const DataFormats* dataFormats, int region) :
+    enableTruncation_(iConfig.getParameter<bool>("EnableTruncation")),
+    setup_(setup),
+    dataFormats_(dataFormats),
+    region_(region),
+    input_(dataFormats_->numChannel(Process::gp), vector<deque<StubPP*>>(dataFormats_->numChannel(Process::pp))) {}
+
+  void GeometricProcessor::consume(const TTDTC& ttDTC) {
+    auto validFrame = [](int& sum, const TTDTC::Frame& frame){ return sum += frame.first.isNonnull() ? 1 : 0; };
+    int nStubsPP(0);
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::pp); channel++) {
+      const TTDTC::Stream& stream = ttDTC.stream(region_, channel);
+      nStubsPP += accumulate(stream.begin(), stream.end(), 0, validFrame);
+    }
+    stubsPP_.reserve(nStubsPP);
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::pp); channel++) {
+      for (const TTDTC::Frame& frame : ttDTC.stream(region_, channel)) {
+        StubPP* stub = nullptr;
+        if (frame.first.isNonnull()) {
+          stubsPP_.emplace_back(frame, dataFormats_);
```

Comment:
Can't we delete stubsPP_? The only place it is used in L34-35, when a stub is put into it and then immediately retrieved again.

Reply:
those contain the input and output stubs. How shall we emulate the GP without stubs?

Reply:
Understood. Please add comment such as "Store input stubs, so remainder of GP algo can work with pointers to them".